src/main/java/net/minecraftforge/registries/holdersets/CompositeHolderSet.java

```diff
@@ -161,4 +162,35 @@ public Iterator<Holder<T>> iterator()
     {
         return this.getList().iterator();
     }
+    
+    /**
+     * Maps the sub-holdersets of this composite such that,if the list contains more than one element,
+     * each element of the list will serialize as an object.
+     * Prevents crashes from trying to serialize non-homogenous lists to NBT.
+     * @return List of holdersets with homogenous serialization behavior.
+     * Returns a new List if size > 1, otherwise returns the composite's existing List.
+     */
+    public List<HolderSet<T>> homogenize()
+    {
+        List<HolderSet<T>> components = this.getComponents();
+        
+        if (components.size() <= 1) // avoid infinitely looping on single-element holdersets
+        {
+            return components;
+        }
+        
+        List<HolderSet<T>> outputs = new ArrayList<>();
+        for (HolderSet<T> holderset : components)
+        {
+            if (holderset instanceof ICustomHolderSet<T>) // serializes to object already
+            {
+                outputs.add(holderset);
+            }
+            else // serializes to string or list
+            {
+                outputs.add(new OrHolderSet<>(List.of(holderset)));
```

Comment:
Would it be worth only wrapping it if there is at least one custom holder set? Something like

```json
[ "#some:tag", "#some:othertag" ]
```
is not problematic afterall.

Reply:
We could but then we'd also have to check if it contains a mix of strings and lists

Direct holderset with one element -> serializes as string
Direct holderset with 0 or 2 or more elements -> serializes as list
Named holderset (tag) -> serializes as string
ICustomHolderset -> serializes as object

or we could just say "don't use list holdersets with tags, just make another tag"